util/sonar.sh

```diff
@@ -2,55 +2,80 @@
-includes="include, test/include, $tmp/include"
-compilerincludes="$(echo "-I$includes" | sed 's/, / -I/g')"
-
-for dir in $(echo | gcc -Wp,-v -x c++ - -fsyntax-only 2>&1 | awk '/#include <...> search starts here:/{f=1;next} /End of search list./{f=0} f')
-do
-	includes="$includes, $dir"
-done
-
-version=$(git rev-parse --short HEAD)
-filename="sonar-project.properties"
-cp util/$filename.template $filename
-echo "sonar.projectVersion=$version" >> $filename
-echo "sonar.cxx.includeDirectories=$includes" >> $filename
-echo "sonar.password=$SONAR_PASSWORD" >> $filename
-echo "sonar.login=$SONAR_ACCESS_TOKEN" >> $filename
-echo "sonar.host.url=$SONAR_URL" >> $filename
-
-(cd $tmp && cmake -DCMAKE_BUILD_TYPE=Debug \
-                  -DCMAKE_CXX_FLAGS="--coverage -Wall -Wextra -fdiagnostics-show-option" \
-                  -DCMAKE_EXE_LINKER_FLAGS="--coverage" \
-                  -DOGDF_SEPARATE_TESTS=ON \
-                  -DOGDF_WARNING_ERRORS=ON ..) || die "Could not run CMake"
-make -C $tmp -j "$cores" COIN 2>/dev/null || die "make COIN failed"
-make -C $tmp -j "$cores" build-all 2>log/gcc.log || die "make build-all failed"
-util/perform_separate_tests.sh $tmp || die "performing separate tests failed"
-gcovr -x -r . --object-directory=$tmp/CMakeFiles >log/gcovr.xml || die "gcovr failed"
-
-sourcefiles="$(git ls-files 'src/ogdf/*.cpp' 'test/src/*.cpp' | check_filter)"
-unusedheaders="$(util/find_unused_headers.sh $tmp | check_filter)"
-cppcheck -j "$cores" --enable=style $compilerincludes --force --xml --language=c++ $sourcefiles $unusedheaders 2>log/cppcheck.xml || die "cppcheck failed"
-
-# point sonar-scanner to certificate so it can communicate with sonarqube server over https
-sonar_keystore="$(echo /sonar-scanner*/jre/lib/security/cacerts)"
-export SONAR_SCANNER_OPTS=-Djavax.net.ssl.trustStore="$sonar_keystore"
-sonar-scanner || die "sonar-runner failed"
+set -e
+mkdir -p build-sonar sonar
+
+git ls-files --full-name -- 'src/ogdf/*.cpp' 'test/src/*.cpp' | check_filter | sort | uniq > sonar/all-sources.txt
+git ls-files --full-name -- 'include/ogdf/*.h' | check_filter | sort | uniq > sonar/all-headers.txt
+
+## generate compile_commands.json
+opts="cmake .. -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_CXX_COMPILER='clang++' "
+
+## debug config
+opts+="-DCMAKE_CXX_FLAGS_DEBUG='-g -O1' "
+opts+="-DBUILD_SHARED_LIBS=ON "
+opts+="-DCMAKE_BUILD_TYPE=Debug "
+opts+="-DOGDF_DEBUG_MODE=HEAVY "
+opts+="-DOGDF_USE_ASSERT_EXCEPTIONS=ON "
+opts+="-DOGDF_MEMORY_MANAGER=POOL_TS "
+
+## include CGAL
+opts+="-DOGDF_INCLUDE_CGAL=ON "
+opts+="-DCGAL_DO_NOT_WARN_ABOUT_CMAKE_BUILD_TYPE=TRUE "
+
+## coverage
+opts+="-DOGDF_SEPARATE_TESTS=ON "
+opts+="-DCMAKE_CXX_FLAGS='--coverage -Wall -Wextra -fdiagnostics-show-option' "
+opts+="-DCMAKE_EXE_LINKER_FLAGS='--coverage' "
+
+## cmd-line args
+opts+="$@"
+
+# Compile!
+echo "::group::($(date -Iseconds)) Compile"
+cd build-sonar
+echo $opts
+eval $opts
+make -j "$cores" build-all
+cp compile_commands.json ..
```

Comment:
I do not understand why we need this.

Reply:
`gcovr` and `clang-tidy` search for `compile_commands.json` in the root dir, while SonarCloud is configured to use `sonar/compile_commands.json`, which is also uploaded as build artifact together with the whole directory. So the second copy is mostly for also having the file included in the artifact zip.

Reply:
Ah, ok, I missed that `gcovr` and `clang-tidy` searched in the root dir, sorry!